internal/cmd/dockerbuild/main.go

```diff
@@ -1,55 +1,166 @@
-	plugins, err := plugin.FindAll(basedir)
+	allPlugins, err := plugin.FindAll(c.pluginsDir)
 	if err != nil {
 		return err
 	}
-	includedPlugins, err := plugin.FilterByPluginsEnv(plugins, os.Getenv("PLUGINS"))
+	includedPlugins, err := plugin.FilterByPluginsEnv(allPlugins, os.Getenv("PLUGINS"))
 	if err != nil {
 		return err
 	}
 	if len(includedPlugins) == 0 {
 		return nil // nothing to build
 	}
-	for _, pluginToBuild := range includedPlugins {
+	pluginGroups, err := getPluginGroups(includedPlugins)
+	if err != nil {
+		return err
+	}
+	// Build bazel plugins on their own as they are very resource intensive.
+	if bazelPlugins, ok := pluginGroups[bazelPluginGroup]; ok {
+		delete(pluginGroups, bazelPluginGroup)
+		// Sort plugins to build first by version, then by name.
+		// This ensures the best use of the Docker build cache for expensive builds like protoc/grpc plugins.
+		slices.SortFunc(bazelPlugins, func(a, b *plugin.Plugin) int {
+			if v := semver.Compare(a.PluginVersion, b.PluginVersion); v != 0 {
+				return v
+			}
+			return cmp.Compare(a.Name, b.Name)
+		})
+		if err := c.buildPluginGroup(ctx, bazelPluginGroup, bazelPlugins); err != nil {
+			return err
+		}
+		if len(pluginGroups) == 0 {
+			return nil
+		}
+	}
+	limit := runtime.GOMAXPROCS(0)
+	if limit > maxLimit {
+		limit = maxLimit
+	}
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	eg.SetLimit(limit)
+	for pluginGroup, plugins := range pluginGroups {
+		pluginGroup := pluginGroup
+		plugins := plugins
+		eg.Go(func() error {
+			return c.buildPluginGroup(ctx, pluginGroup, plugins)
+		})
+	}
+	return eg.Wait()
+}
+
+func getPluginGroups(plugins []*plugin.Plugin) (map[string][]*plugin.Plugin, error) {
+	pluginGroups := make(map[string][]*plugin.Plugin)
+	for _, pluginToBuild := range plugins {
+		var pluginKey string
+		identity := pluginToBuild.Identity
+		// Group grpc/protobuf builds together so one finishes to completion before running additional jobs.
+		// This is important because builds can share the Docker cache to optimize longer Bazel builds.
+		switch owner := identity.Owner(); owner {
+		case "grpc":
+			switch identity.Plugin() {
+			case "cpp", "csharp", "objc", "php", "python", "ruby":
+				pluginKey = bazelPluginGroup
+			}
+		case "protocolbuffers":
+			switch identity.Plugin() {
+			case "cpp", "csharp", "java", "kotlin", "objc", "php", "pyi", "python", "ruby", "rust":
+				pluginKey = bazelPluginGroup
+			}
+		default:
+			// Assume everything else can be built independently
+			pluginKey = identity.IdentityString()
+		}
+		pluginGroups[pluginKey] = append(pluginGroups[pluginKey], pluginToBuild)
+	}
+	return pluginGroups, nil
+}
+
+func (c *command) buildPluginGroup(ctx context.Context, pluginGroup string, plugins []*plugin.Plugin) error {
+	for _, pluginToBuild := range plugins {
+		pluginIdentity := pluginToBuild.Identity
+		var pluginCacheDir string
+		// If cache is enabled, create a unique cache directory per group of plugins.
```

Comment:
This is the new logic - we will create a unique cache directory for each plugin group. Here's an example:

```
.tmp/dockercache
├── bazel
│   ├── grpc
│   │   └── v1.58.1
│   └── protocolbuffers
│       └── v24.3
├── connectrpc
│   └── go
│       └── v1.11.1
└── grpc
    └── java
        └── v1.58.0
```